File: javatests/com/google/protobuf/contrib/j2cl/integration/LongFieldsTest.java
/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package com.google.protobuf.contrib.j2cl.integration;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.Assert.assertThrows;

import com.google.protobuf.contrib.j2cl.protos.Accessor.TestProto;
import java.util.Arrays;
import java.util.List;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class LongFieldsTest {

  private static final long DEFAULT_PROTO_VALUE = 3000000000L;

  @Test
  public void testOptionalFieldNoDefault_defaultInstance() {
    assertThat(TestProto.getDefaultInstance().hasOptionalLong()).isFalse();
    assertThat(TestProto.getDefaultInstance().getOptionalLong()).isEqualTo(0L);
  }

  @Test
  public void testOptionalFieldNoDefault_setValue() {
    TestProto.Builder builder = TestProto.newBuilder().setOptionalLong(8964L);
    assertThat(builder.hasOptionalLong()).isTrue();
    assertThat(builder.getOptionalLong()).isEqualTo(8964L);

    TestProto proto = builder.build();
    assertThat(proto.hasOptionalLong()).isTrue();
    assertThat(proto.getOptionalLong()).isEqualTo(8964L);
  }

  @Test
  public void testOptionalFieldNoDefault_setDefaultValue() {
    TestProto.Builder builder = TestProto.newBuilder().setOptionalLong(0L);
    assertThat(builder.hasOptionalLong()).isTrue();
    assertThat(builder.getOptionalLong()).isEqualTo(0L);

    TestProto proto = builder.build();
    assertThat(proto.hasOptionalLong()).isTrue();
    assertThat(proto.getOptionalLong()).isEqualTo(0L);
  }

  @Test
  public void testOptionalFieldNoDefault_clear() {
    TestProto startProto = TestProto.newBuilder().setOptionalLong(0L).build();

    TestProto.Builder builder = startProto.toBuilder();
    builder.clearOptionalLong();
    assertThat(builder.hasOptionalLong()).isFalse();

    TestProto proto = builder.build();
    assertThat(proto.hasOptionalLong()).isFalse();
  }

  @Test
  public void testOptionalFieldWithDefault_defaultInstance() {
    assertThat(TestProto.getDefaultInstance().hasOptionalLongWithDefault()).isFalse();
    assertThat(TestProto.getDefaultInstance().getOptionalLongWithDefault())
        .isEqualTo(DEFAULT_PROTO_VALUE);
  }

  @Test
  public void testOptionalFieldWithDefault_setValue() {
    TestProto.Builder builder = TestProto.newBuilder().setOptionalLongWithDefault(-4000000000L);
    assertThat(builder.hasOptionalLongWithDefault()).isTrue();
    assertThat(builder.getOptionalLongWithDefault()).isEqualTo(-4000000000L);

    TestProto proto = builder.build();
    assertThat(proto.hasOptionalLongWithDefault()).isTrue();
    assertThat(proto.getOptionalLongWithDefault()).isEqualTo(-4000000000L);
  }

  @Test
  public void testOptionalFieldWithDefault_setDefaultValue() {
    TestProto.Builder builder =
        TestProto.newBuilder().setOptionalLongWithDefault(DEFAULT_PROTO_VALUE);
    assertThat(builder.hasOptionalLongWithDefault()).isTrue();
    assertThat(builder.getOptionalLongWithDefault()).isEqualTo(DEFAULT_PROTO_VALUE);

    TestProto proto = builder.build();
    assertThat(proto.hasOptionalLongWithDefault()).isTrue();
    assertThat(proto.getOptionalLongWithDefault()).isEqualTo(DEFAULT_PROTO_VALUE);
  }

  @Test
  public void testRepeatedField_defaultInstance() {
    assertThat(TestProto.getDefaultInstance().getRepeatedLongCount()).isEqualTo(0);
    assertThat(TestProto.newBuilder().build().getRepeatedLongCount()).isEqualTo(0);
    assertThrows(
        Exception.class,
        () -> {
          @SuppressWarnings("UnusedVariable")
          Long willFailToBox = TestProto.newBuilder().build().getRepeatedLong(0);
        });
  }

  @Test
  public void testRepeatedField_add() {
    TestProto.Builder builder =
        TestProto.newBuilder().addRepeatedLong(13L).addRepeatedLong(24000000000L);
    assertThat(Arrays.asList(13L, 24000000000L))
        .containsExactlyElementsIn(builder.getRepeatedLongList())
        .inOrder();

    TestProto proto = builder.build();

    assertThat(Arrays.asList(13L, 24000000000L))
        .containsExactlyElementsIn(proto.getRepeatedLongList())
        .inOrder();
  }

  @Test
  public void testRepeatedField_addAll() {
    TestProto startProto =
        TestProto.newBuilder().addRepeatedLong(13L).addRepeatedLong(24000000000L).build();

    TestProto.Builder builder =
        startProto.toBuilder().addAllRepeatedLong(Arrays.asList(-1L, -2L, -3L));

    assertThat(Arrays.asList(13L, 24000000000L, -1L, -2L, -3L))
        .containsExactlyElementsIn(builder.getRepeatedLongList())
        .inOrder();

    TestProto proto = builder.build();
    assertThat(Arrays.asList(13L, 24000000000L, -1L, -2L, -3L))
        .containsExactlyElementsIn(proto.getRepeatedLongList())
        .inOrder();
  }

  @Test
  public void testRepeatedField_set() {
    TestProto startProto =
        TestProto.newBuilder()
            .addRepeatedLong(13L)
            .addRepeatedLong(24000000000L)
            .addRepeatedLong(-1L)
            .addRepeatedLong(-2L)
            .addRepeatedLong(-3L)
            .build();

    TestProto.Builder builder = startProto.toBuilder().setRepeatedLong(2, 333L);
    assertThat(Arrays.asList(13L, 24000000000L, 333L, -2L, -3L))
        .containsExactlyElementsIn(builder.getRepeatedLongList())
        .inOrder();

    TestProto proto = builder.build();
    assertThat(Arrays.asList(13L, 24000000000L, 333L, -2L, -3L))
        .containsExactlyElementsIn(proto.getRepeatedLongList())
        .inOrder();
  }

  @Test
  public void testRepeatedField_getAndCount() {
    TestProto.Builder builder =
        TestProto.newBuilder()
            .addRepeatedLong(13L)
            .addRepeatedLong(24000000000L)
            .addRepeatedLong(-1L)
            .addRepeatedLong(-2L)
            .addRepeatedLong(-3L);

    assertThat(builder.getRepeatedLongCount()).isEqualTo(5);
    assertThat(builder.getRepeatedLong(0)).isEqualTo(13L);
    assertThat(builder.getRepeatedLong(1)).isEqualTo(24000000000L);
    assertThat(builder.getRepeatedLong(2)).isEqualTo(-1L);
    assertThat(builder.getRepeatedLong(3)).isEqualTo(-2L);
    assertThat(builder.getRepeatedLong(4)).isEqualTo(-3L);
    assertThrows(
        Exception.class,
        () -> {
          @SuppressWarnings("UnusedVariable")
          Long willFailToBox = builder.getRepeatedLong(5);
        });

    TestProto proto = builder.build();
    assertThat(proto.getRepeatedLongCount()).isEqualTo(5);
    assertThat(proto.getRepeatedLong(0)).isEqualTo(13L);
    assertThat(proto.getRepeatedLong(1)).isEqualTo(24000000000L);
    assertThat(proto.getRepeatedLong(2)).isEqualTo(-1L);
    assertThat(proto.getRepeatedLong(3)).isEqualTo(-2L);
    assertThat(proto.getRepeatedLong(4)).isEqualTo(-3L);
    assertThrows(
        Exception.class,
        () -> {
          @SuppressWarnings("UnusedVariable")
          Long willFailToBox = proto.getRepeatedLong(5);
        });
  }

  @Test
  public void testRepeatedField_clear() {
    TestProto startProto =
        TestProto.newBuilder()
            .addRepeatedLong(13L)
            .addRepeatedLong(24000000000L)
            .addRepeatedLong(-1L)
            .addRepeatedLong(-2L)
            .addRepeatedLong(-3L)
            .build();
    TestProto.Builder builder = startProto.toBuilder().clearRepeatedLong();
    assertThat(builder.getRepeatedLongCount()).isEqualTo(0);
    assertThat(builder.getRepeatedLongList()).isEmpty();

    TestProto proto = builder.build();
    assertThat(proto.getRepeatedLongCount()).isEqualTo(0);
    assertThat(proto.getRepeatedLongCount()).isEqualTo(0);
  }

  @Test
  public void testRepeatedField_getReturnsImmutableList() {
    TestProto.Builder builder = TestProto.newBuilder().addRepeatedLong(1);
    List<Long> repeatedFieldList = builder.getRepeatedLongList();

    assertThrows(Exception.class, () -> repeatedFieldList.add(2L));
    assertThrows(Exception.class, () -> repeatedFieldList.remove(0));
  }
}

File: javatests/com/google/protobuf/contrib/immutablejs/integration/long_fields_test.js
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

goog.module('proto.im.integration.LongFieldsTest');
goog.setTestOnly();


const ListView = goog.require('proto.im.ListView');
const Long = goog.require('goog.math.Long');
const TestProto = goog.require('improto.protobuf.contrib.immutablejs.protos.TestProto');
const testSuite = goog.require('goog.testing.testSuite');
const {assertEqualsForProto} = goog.require('proto.im.proto_asserts');
const {isCheckIndex} = goog.require('proto.im.internal.internalChecks');


const DEFAULT_PROTO_VALUE = Long.fromString('3000000000');

class LongFieldsTest {
  testOptionalFieldNoDefault_defaultInstance() {
    assertEqualsForProto(
        false, TestProto.newBuilder().build().hasOptionalLong());
    assertEqualsForProto(
        Long.fromInt(0), TestProto.newBuilder().build().getOptionalLong());
  }

  testOptionalFieldNoDefault_setValue() {
    const builder = TestProto.newBuilder().setOptionalLong(Long.fromInt(8964));
    assertEqualsForProto(true, builder.hasOptionalLong());
    assertEqualsForProto(Long.fromInt(8964), builder.getOptionalLong());

    const proto = builder.build();
    assertEqualsForProto(true, proto.hasOptionalLong());
    assertEqualsForProto(Long.fromInt(8964), proto.getOptionalLong());
  }

  testOptionalFieldNoDefault_setDefaultValue() {
    const builder = TestProto.newBuilder().setOptionalLong(Long.fromInt(0));
    assertEqualsForProto(true, builder.hasOptionalLong());
    assertEqualsForProto(Long.fromInt(0), builder.getOptionalLong());

    const proto = builder.build();
    assertEqualsForProto(true, proto.hasOptionalLong());
    assertEqualsForProto(Long.fromInt(0), proto.getOptionalLong());
  }

  testOptionalFieldNoDefault_clear() {
    const startProto =
        TestProto.newBuilder().setOptionalLong(Long.fromInt(0)).build();
    // Test clear
    const builder = startProto.toBuilder();
    builder.clearOptionalLong();
    assertEqualsForProto(false, builder.hasOptionalLong());

    const proto = builder.build();
    assertEqualsForProto(false, proto.hasOptionalLong());
  }

  testOptionalFieldWithDefault_defaultInstance() {
    assertEqualsForProto(
        false, TestProto.newBuilder().build().hasOptionalLongWithDefault());
    assertEqualsForProto(
        DEFAULT_PROTO_VALUE,
        TestProto.newBuilder().build().getOptionalLongWithDefault());
  }
  testOptionalFieldWithDefault_setValue() {
    const builder = TestProto.newBuilder().setOptionalLongWithDefault(
        Long.fromString('-4000000000'));
    assertEqualsForProto(true, builder.hasOptionalLongWithDefault());
    assertEqualsForProto(
        Long.fromString('-4000000000'), builder.getOptionalLongWithDefault());

    const proto = builder.build();
    assertEqualsForProto(true, proto.hasOptionalLongWithDefault());
    assertEqualsForProto(
        Long.fromString('-4000000000'), proto.getOptionalLongWithDefault());
  }

  testOptionalFieldWithDefault_setDefaultValue() {
    const builder =
        TestProto.newBuilder().setOptionalLongWithDefault(DEFAULT_PROTO_VALUE);
    assertEqualsForProto(true, builder.hasOptionalLongWithDefault());
    assertEqualsForProto(
        DEFAULT_PROTO_VALUE, builder.getOptionalLongWithDefault());

    const proto = builder.build();
    assertEqualsForProto(true, proto.hasOptionalLongWithDefault());
    assertEqualsForProto(
        DEFAULT_PROTO_VALUE, proto.getOptionalLongWithDefault());
  }

  testRepeatedField_defaultInstance() {
    assertEqualsForProto(
        0, TestProto.newBuilder().build().getRepeatedLongCount());
    assertEqualsForProto(
        0, TestProto.newBuilder().build().getRepeatedLongList().size());
    if (isCheckIndex()) {
      assertThrows(() => TestProto.newBuilder().build().getRepeatedLong(0));
    } else {
      assertUndefined(TestProto.newBuilder().build().getRepeatedLong(0));
    }
  }

  testRepeatedField_add() {
    const builder = TestProto.newBuilder()
                        .addRepeatedLong(Long.fromString('13'))
                        .addRepeatedLong(Long.fromString('24000000000'));
    assertEqualsForProto(
        [Long.fromString('13'), Long.fromString('24000000000')],
        builder.getRepeatedLongList().toArray());

    const proto = builder.build();
    assertEqualsForProto(
        [Long.fromString('13'), Long.fromString('24000000000')],
        proto.getRepeatedLongList().toArray());
  }

  testRepeatedField_addAll() {
    const startProto = TestProto.newBuilder()
                           .addRepeatedLong(Long.fromString('13'))
                           .addRepeatedLong(Long.fromString('24000000000'))
                           .build();

    const builder = startProto.toBuilder().addAllRepeatedLong(ListView.copyOf(
        [Long.fromString('-1'), Long.fromString('-2'), Long.fromString('-3')]));
    assertEqualsForProto(
        [
          Long.fromString('13'), Long.fromString('24000000000'),
          Long.fromString('-1'), Long.fromString('-2'), Long.fromString('-3')
        ],
        builder.getRepeatedLongList().toArray());

    const proto = builder.build();
    assertEqualsForProto(
        [
          Long.fromString('13'), Long.fromString('24000000000'),
          Long.fromString('-1'), Long.fromString('-2'), Long.fromString('-3')
        ],
        proto.getRepeatedLongList().toArray());
  }

  testRepeatedField_set() {
    const startProto = TestProto.newBuilder()
                           .addRepeatedLong(Long.fromString('13'))
                           .addRepeatedLong(Long.fromString('24000000000'))
                           .addRepeatedLong(Long.fromString('-1'))
                           .addRepeatedLong(Long.fromString('-2'))
                           .addRepeatedLong(Long.fromString('-3'))
                           .build();

    const builder =
        startProto.toBuilder().setRepeatedLong(2, Long.fromString('333'));
    assertEqualsForProto(
        [
          Long.fromString('13'), Long.fromString('24000000000'),
          Long.fromString('333'), Long.fromString('-2'), Long.fromString('-3')
        ],
        builder.getRepeatedLongList().toArray());

    const proto = builder.build();
    assertEqualsForProto(
        [
          Long.fromString('13'), Long.fromString('24000000000'),
          Long.fromString('333'), Long.fromString('-2'), Long.fromString('-3')
        ],
        proto.getRepeatedLongList().toArray());
  }

  testRepeatedField_getAndCount() {
    const builder = TestProto.newBuilder()
                        .addRepeatedLong(Long.fromString('13'))
                        .addRepeatedLong(Long.fromString('24000000000'))
                        .addRepeatedLong(Long.fromString('-1'))
                        .addRepeatedLong(Long.fromString('-2'))
                        .addRepeatedLong(Long.fromString('-3'));

    assertEqualsForProto(5, builder.getRepeatedLongCount());
    assertEqualsForProto(Long.fromString('13'), builder.getRepeatedLong(0));
    assertEqualsForProto(
        Long.fromString('24000000000'), builder.getRepeatedLong(1));
    assertEqualsForProto(Long.fromString('-1'), builder.getRepeatedLong(2));
    assertEqualsForProto(Long.fromString('-2'), builder.getRepeatedLong(3));
    assertEqualsForProto(Long.fromString('-3'), builder.getRepeatedLong(4));
    if (isCheckIndex()) {
      assertThrows(() => builder.getRepeatedLong(5));
    } else {
      assertUndefined(builder.getRepeatedLong(5));
    }

    const proto = builder.build();
    assertEquals(5, proto.getRepeatedLongCount());
    assertEqualsForProto(
        Long.fromString('24000000000'), proto.getRepeatedLong(1));
    assertEqualsForProto(Long.fromString('-1'), proto.getRepeatedLong(2));
    assertEqualsForProto(Long.fromString('-2'), proto.getRepeatedLong(3));
    assertEqualsForProto(Long.fromString('-3'), proto.getRepeatedLong(4));
    if (isCheckIndex()) {
      assertThrows(() => proto.getRepeatedLong(5));
    } else {
      assertUndefined(proto.getRepeatedLong(5));
    }
  }

  testRepeatedField_clear() {
    const startProto = TestProto.newBuilder()
                           .addRepeatedLong(Long.fromString('13'))
                           .addRepeatedLong(Long.fromString('24000000000'))
                           .addRepeatedLong(Long.fromString('-1'))
                           .addRepeatedLong(Long.fromString('-2'))
                           .addRepeatedLong(Long.fromString('-3'))
                           .build();
    const builder = startProto.toBuilder().clearRepeatedLong();
    assertEqualsForProto(0, builder.getRepeatedLongCount());
    assertEqualsForProto(0, builder.getRepeatedLongList().size());


    const proto = builder.build();
    assertEqualsForProto(0, proto.getRepeatedLongCount());
    assertEqualsForProto(0, proto.getRepeatedLongList().size());
  }

  testNoDataLoss() {
    const builder = TestProto.newBuilder().setOptionalLongWithDefault(
        Long.fromNumber(Math.pow(2, 63)));
    assertEqualsForProto(
        Long.fromNumber(Math.pow(2, 63)), builder.getOptionalLongWithDefault());

    builder.setOptionalLongWithDefault(Long.fromNumber(-Math.pow(2, 63)));
    assertEqualsForProto(
        Long.fromNumber(-Math.pow(2, 63)),
        builder.getOptionalLongWithDefault());
  }
}

testSuite(new LongFieldsTest());
